state-chain/pallets/cf-asset-balances/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+		available: &mut AssetAmount,
+	) -> Result<(), DispatchError> {
+		if *amount_owed > *available {
+			Self::deposit_event(Event::VaultDeficitDetected {
+				chain,
+				amount_owed: *amount_owed,
+				available: *available,
+			});
+		}
+		let amount_reconciled = match chain {
+			ForeignChain::Ethereum | ForeignChain::Arbitrum => match owner {
+				ExternalOwner::Account(address) =>
+					if *amount_owed > *available {
+						0
+					} else {
+						Self::refund_via_egress(chain, address.clone(), *amount_owed)?;
+						*amount_owed
+					},
+				other => {
+					log_or_panic!(
+						"{:?} Liabilities are not supported for chain {:?}.",
+						other,
+						chain
+					);
+					0
+				},
+			},
+			ForeignChain::Polkadot => {
+				match owner {
+					ExternalOwner::AggKey => {
+						let address = ForeignChainAddress::Dot(
+							T::PolkadotKeyProvider::active_epoch_key()
+								// TODO: make this more defensive
+								.expect("Key must exist otherwise we would have nothing to refund.")
+								.key,
+						);
+						let refund_amount = core::cmp::min(*amount_owed, *available);
+						Self::refund_via_egress(chain, address, refund_amount)?;
+						refund_amount
+					},
+					other => {
+						log_or_panic!(
+							"{:?} Liabilities are not supported for chain {:?}.",
+							other,
+							chain
+						);
+						0
+					},
+				}
+			},
+			ForeignChain::Bitcoin | ForeignChain::Solana => match owner {
+				ExternalOwner::Vault => core::cmp::min(*amount_owed, *available),
+				other => {
+					log_or_panic!(
+						"{:?} Liabilities are not supported for chain {:?}.",
+						other,
+						chain
+					);
+					0
+				},
+			},
+		};
+
+		available.defensive_saturating_reduce(amount_reconciled);
+		amount_owed.defensive_saturating_reduce(amount_reconciled);
+
+		Ok(())
+	}
+
+	pub fn trigger_reconciliation() {
+		if !T::SafeMode::get().reconciliation_enabled {
+			log::info!("Reconciliation is disabled. Skipping reconciliation.");
+			return;
+		}
+
+		for chain in ForeignChain::iter() {
+			let owed_assets = Liabilities::<T>::take(chain.gas_asset());
+			WithheldAssets::<T>::mutate(chain.gas_asset(), |total_withheld| {
+				let mut owed_assets = owed_assets.into_iter().collect::<Vec<_>>();
+				owed_assets.sort_by_key(|(_, amount)| core::cmp::Reverse(*amount));
+
+				for (destination, amount) in &mut owed_assets {
+					debug_assert!(*amount > 0);
+					match Self::reconcile(chain, destination, amount, total_withheld) {
+						Ok(_) =>
+							if *total_withheld == 0 {
+								break;
+							},
+						Err(_) => break,
+					}
+				}
+
+				owed_assets.retain(|(_, amount)| *amount > 0);
+				if !owed_assets.is_empty() {
+					Liabilities::<T>::insert(
+						chain.gas_asset(),
+						owed_assets.into_iter().collect::<BTreeMap<_, _>>(),
+					);
+				}
+			});
+		}
+	}
+
+	pub fn vault_imbalance(asset: Asset) -> VaultImbalance<AssetAmount> {
+		let owed = Liabilities::<T>::get(asset).values().sum::<u128>();
```

Comment:
I think it's the same reason again... We just need this for the migration and then want to remove it after the migration is done. Most likely we will end up with an inline solution again then, because we argue that it's only used a single time inside the pallet itself and there is no point in having it implemented in the trait or even in an own method.

Reply:
Ok, we should create a ticket to not forget this then, can label it with "1.6" so we address it before 1.6 is released.

Reply:
https://linear.app/chainflip/issue/PRO-1515/remove-migration-feature-gated-functions-after-15